cura/PrinterOutputDevice.py

```diff
@@ -0,0 +1,303 @@
+        Logger.log("w", "_homeBed is not implemented by this output device")
+
+    ##  Protected setter for the bed temperature of the connected printer (if any).
+    #   /parameter temperature Temperature bed needs to go to (in deg celsius)
+    #   /sa setTargetBedTemperature
+    def _setTargetBedTemperature(self, temperature):
+        Logger.log("w", "_setTargetBedTemperature is not implemented by this output device")
+
+    ##  Protected setter for the current bed temperature.
+    #   This simply sets the bed temperature, but ensures that a signal is emitted.
+    #   /param temperature temperature of the bed.
+    def _setBedTemperature(self, temperature):
+        self._bed_temperature = temperature
+        self.bedTemperatureChanged.emit()
+
+    ##  Get the bed temperature if connected printer (if any)
+    @pyqtProperty(int, notify = bedTemperatureChanged)
+    def bedTemperature(self):
+        return self._bed_temperature
+
+    ##  Get the target bed temperature if connected printer (if any)
+    @pyqtProperty(int, notify = targetBedTemperatureChanged)
+    def targetBedTemperature(self):
+        return self._target_bed_temperature
+
+    ##  Set the (target) hotend temperature
+    #   This function is "final" (do not re-implement)
+    #   /param index the index of the hotend that needs to change temperature
+    #   /param temperature The temperature it needs to change to (in deg celsius).
+    #   /sa _setTargetHotendTemperature implementation function
+    @pyqtSlot(int, int)
+    def setTargetHotendTemperature(self, index, temperature):
+        self._setTargetHotendTemperature(index, temperature)
+        self._target_hotend_temperatures[index] = temperature
+        self.targetHotendTemperaturesChanged.emit()
+
+    ##  Implementation function of setTargetHotendTemperature.
+    #   /param index Index of the hotend to set the temperature of
+    #   /param temperature Temperature to set the hotend to (in deg C)
+    #   /sa setTargetHotendTemperature
+    def _setTargetHotendTemperature(self, index, temperature):
+        Logger.log("w", "_setTargetHotendTemperature is not implemented by this output device")
+
+    @pyqtProperty("QVariantList", notify = targetHotendTemperaturesChanged)
+    def targetHotendTemperatures(self):
+        return self._target_hotend_temperatures
+
+    @pyqtProperty("QVariantList", notify = hotendTemperaturesChanged)
+    def hotendTemperatures(self):
+        return self._hotend_temperatures
+
+    ##  Protected setter for the current hotend temperature.
+    #   This simply sets the hotend temperature, but ensures that a signal is emitted.
+    #   /param index Index of the hotend
+    #   /param temperature temperature of the hotend (in deg C)
+    def _setHotendTemperature(self, index, temperature):
+        self._hotend_temperatures[index] = temperature
+        self.hotendTemperaturesChanged.emit()
+
+    ##  Attempt to establish connection
+    def connect(self):
+        pass
```

Comment:
Shouldn't this raise NotImplementedError? Seems both this and the other should always be reimplemented.